Node Tutorial Guide (#68)
celestiaorg/docs: docs/developers/node-tutorial.md

In this tutorial we will cover how to use the Celestia Node API to submit and
retrieve transactions from the Data Availability Layer by their namespace
merkle tree.

## Hardware Requirements

You can find hardware requirements [here](https://docs.celestia.org/nodes/bridge-validator-node#install-golang).

## Setting Up Dependencies

You can follow the tutorial for setting up the dependencies [here](https://docs.celestia.org/nodes/bridge-validator-node#setting-up-your-bridge-node).

## Installing Celestia App

In order to begin, we would need to compile the Celestia App binary.

```sh
cd $HOME
git clone https://github.com/celestiaorg/celestia-app.git
cd celestia-app/
git checkout tags/v0.4.0 -b v0.4.0
make install
```

For more information, you can follow the steps on compiling the binary [here](https://docs.celestia.org/nodes/bridge-validator-node#install-celestia-app).

Once `celestia-appd` is compiled, you can run a useful script for instantiating
a single-node network.

In the `contrib` directory, open up the file `single-node.sh` and replace its
content with the following script:

```sh
#!/bin/sh

set -o errexit -o nounset

CHAINID="devnet-2"

> app="./celestia-appd"

Reviewer: are we using the directory as a binary here?

Reviewer 2: Likely a relic from my own script.

Reviewer 2: They should have $app point to the binary or take it as an arg in the script.

# Build genesis file incl account for passed address
coins="1000000000000000uceles"
$app init $CHAINID --chain-id $CHAINID 
$app keys add validator --keyring-backend="test"
$app keys add celes --keyring-backend="test"
# this won't work because the some proto types are decalared twice and 
# the logs output to stdout (dependency hell involving iavl)
$app add-genesis-account $($app keys show validator -a --keyring-backend="test") $coins
$app add-genesis-account $($app keys show celes -a --keyring-backend="test") $coins
$app gentx validator 5000000000uceles --keyring-backend="test" --chain-id $CHAINID
$app collect-gentxs

# Set proper defaults and change ports
sed 's#"tcp://127.0.0.1:26657"#"tcp://0.0.0.0:26657"#g' ~/.celestia-app/config/config.toml
sed 's/timeout_commit = "5s"/timeout_commit = "1s"/g' ~/.celestia-app/config/config.toml
sed 's/timeout_propose = "3s"/timeout_propose = "1s"/g' ~/.celestia-app/config/config.toml

> sed 's/index_all_keys = false/index_all_keys = true/g' ~/.celestia-app/config/config.toml

Reviewer 3: assuming we are running latest and greatest, we should include the sed changing from default "full" to "validator" 
more info here: https://github.com/celestiaorg/celestia-app/issues/422

# Start the celestia-app
$app start --grpc.enable
```

Next, run the following from the root-directory of celestia-app:

```sh
cp build/celestia-appd contrib/
cd contrib
./single-node.sh
```

This will instantiate the single-node.sh script which will run a single-node
network for celestia-app.

## Celestia Node

In another terminal window, run the following:

```sh
cd $HOME
git clone https://github.com/celestiaorg/celestia-node.git
cd celestia-node/
make install
```

Verify its installed by running the following command:

> ```sh

Reviewer 4: ````suggestion
```console
````

$ celestia version
Semantic version: v0.2.0
Commit: 1fcf0c0bb5d5a4e18b51cf12440ce86a84cf7a72
Build Date: Fri 04 Mar 2022 01:15:07 AM CET
System version: amd64/linux
Golang version: go1.17.5
```

Now, let's instantiate a celestia-light node:

```sh
./build/celestia light init
```

With the light-node installed, we can proceed to copying over the keys from
the celestia-app to the celestia-node so they share the same key.

```sh
mkdir ~/.celestia-light/keys/keyring-test
cp ~/.celestia-app/keyring-test/* ~/.celestia-light/keys/keyring-test
```

This will copy over the keys to your Celestia Light node.

Let's now run the Celestia Light node with a GRPC connection to the running
Celestia App instance:

```sh
./build/celestia light start --core.grpc 127.0.0.1:9090
```

## Node API Calls

With both Celestia App and Celestia Node running in separate terminals, we can
now open a new terminal session to start making API calls!

First, let's find our wallet public addresses generated by
the `single-node.sh` script.

Run the following (Note: You don't have to use VIM to open the file, it is
used here for demonstration)

```sh
vim ~/.celestia-app/config/genesis.json

> ```

Reviewer 4: Most online instructions in see about editing files use `nano`. Maybe use that instead? Ships by default with Ubuntu.

You can request funds to your wallet address using the following command in Discord:

```console
$request <Wallet-Address>
```

Where `<Wallet-Address>` is the `celes1******` address generated
when you created the wallet.

With your wallet funded, you can move on to the next step.

### Connect To A Public Core Endpoint

Let's now run the Celestia Light node with a GRPC connection to
an example public Core Endpoint.

> Note: You are also encouraged to find a community-run API endpoint
  and there are several in the Discord. This one is used for demonstration
  purposes.

Here we are starting a light node with a connection to a Core endpoint at
64.227.74.87:9090 and also telling the light node to use the `developer`
key we generated as its default account.

```sh
./build/celestia light start --core.grpc 64.227.74.87:9090 --keyring.accname developer
```

## Node API Calls

Open up another terminal window in order to begin querying the API.
`celestia-node` exposes its RPC endpoint on port `26658` by default.

### Balance

Now, let's query our node for the balance of its default account
(which is the account associated with the `developer` key we generated earlier):

```sh

> curl -X GET http://127.0.0.1:26658/balance/

Author: @renaynay don't we need to include the celes address here as in `/balance/<ADDRESS>`

> curl -X GET http://127.0.0.1:26658/balance/

Author: @renaynay don't we need to include the celes address here as in `/balance/<ADDRESS>`

### Generate A Wallet

Once the binary is compiled, you can use it to generate a wallet.

Run the following:

```sh
celestia-appd keys add developer --keyring-backend test
```

This generates a key called `developer` in `celestia-appd`.

Now, head over to the Celestia Discord channel `#faucet`.

You can request funds to your wallet address using the following command in Discord:

```console
$request <Wallet-Address>
```

Where `<Wallet-Address>` is the `celes1******` address generated
when you created the wallet.

With your wallet funded, you can move on to the next step.

## Celestia Node

In another terminal window, run the following:

```sh
cd $HOME
git clone https://github.com/celestiaorg/celestia-node.git
cd celestia-node/
make install
```

Verify its installed by running the following command:

```console
$ celestia version
Semantic version: v0.2.0
Commit: 1fcf0c0bb5d5a4e18b51cf12440ce86a84cf7a72
Build Date: Fri 04 Mar 2022 01:15:07 AM CET
System version: amd64/linux
Golang version: go1.17.5
```

Now, let's instantiate a celestia-light node:

```sh
./build/celestia light init
```

> 

Reviewer 2: ```suggestion
### Generate A Wallet
Once the binary is compiled, you can use it to generate a wallet.

Run the following:
`celestia light keys add developer --keyring-backend test`

This generates a key called `developer` in `~/.celestia-light/keys/keyring-test`.

Now, head over to the Celestia Discord channel `#faucet`.

You can request funds to your wallet address using the following command in Discord:

`$request <Wallet-Address>`

Where `<Wallet-Address>` is the `celes1******` address generated when you created the wallet.

With your wallet funded, you can move on to the next step.
```

Pls format this section to md^

> With the light-node installed, we can proceed to copying over the keys from
> the celestia-app to the celestia-node so they share the same key.
> 
> ```sh
> mkdir ~/.celestia-light/keys/keyring-test
> cp ~/.celestia-app/keyring-test/* ~/.celestia-light/keys/keyring-test
> ```
> 
> This will copy over the keys to your Celestia Light node.
> 

Reviewer 2: ```suggestion
```

Let's now run the Celestia Light node with a GRPC connection to
an example public Core Endpoint.

> Note: You are also encouraged to find a community-run API endpoint
  and there are several in the Discord. This one is used for demonstration
  purposes.

> ```sh
> ./build/celestia light start --core.grpc http://64.227.74.87:26657/
> ```

Reviewer 2: ```suggestion

Here we are starting a light node with a connection to a Core endpoint at 64.227.74.87:9090 and also telling the light node to use the `developer` key we generated as its default account.

```sh
./build/celestia light start --core.grpc 64.227.74.87:9090 --keyring.accname developer
```

Btw: please also note that grpc is usually exposed on port 9090 on celestia-core/app

> ## Node API Calls
> 

Reviewer 2: ```suggestion
## Node API Calls

Celestia-node exposes its RPC endpoint on port `26658` by default.
```

### Balance

> Now, let's query our wallet for its balance:

Reviewer 2: ```suggestion
Now, let's query our node for the balance of its default account (which is the account associated with the `developer` key we generated earlier):
```

```sh

> curl -X GET http://127.0.0.1:26658/balance/celes1v7qgpfwt7jxrhnqj943x584zrhk5wa2qwk67k0

Reviewer 2: ```suggestion
curl -X GET http://127.0.0.1:26658/balance
```

            "voting_power":5000,
            "proposer_priority":5000
         },
         {
            "address":"D345D62BBD18C301B843DF7C65F10E57AB17BD98",
            "pub_key":"4g3hhdyU4IIgWW/4sR0nax8bsC/M/fDbt1N8s/QanF8=",
            "voting_power":5000,
            "proposer_priority":5000
         },
         {
            "address":"DEC2642E786A941511A401090D21621E7F08A36D",
            "pub_key":"b+Vv6Lcp0bhIjOQncr+OYBHixCvU5+k34y4RqyvpluE=",
            "voting_power":5000,
            "proposer_priority":5000
         }
      ],
      "proposer":{
         "address":"03F1044A6DF782189C7061FF89146B3D33608F17",
         "pub_key":"sMcFgSIzlD77eZYgV7H4akyxoHCPc2oIQW05qWEB6b4=",
         "voting_power":5000,
         "proposer_priority":-40000
      }
   },
   "dah":{
      "row_roots":[
         "//////////7//////////uyLCVMJmAItYqbOqgHXm3OwHsq1xQiAX1kZV2Tgcobm",
         "/////////////////////ykyWNfDJZfigziZC5BN5L00KKuoyDPduwynDywauskL"
      ],
      "column_roots":[
         "//////////7//////////uyLCVMJmAItYqbOqgHXm3OwHsq1xQiAX1kZV2Tgcobm",
         "/////////////////////ykyWNfDJZfigziZC5BN5L00KKuoyDPduwynDywauskL"
      ]
   }
}
```

### Submit a PFD Transaction

> In this example, e will be submitting a transaction to the PFD endpoint.

Reviewer 2: ```suggestion
In this example, we will be submitting a PayForData transaction to the node's `/submit_pfd` endpoint.
```

> PFD is a PayForData Message.

Reviewer 2: we should probably include more info about what a PayForData is (maybe not in this doc, but link it to something that explains well what this tx type is bc it's a core part of celestia and not very easy to understand on the outset).

The endpoint also takes in a `namespace_id` and `data` values.

Namespace ID should be 8 bytes.

> Data is in hex-encoded bytes of the raw message.

Reviewer 2: also include description of gas_limit

We use the following `namespace_id` of `0000010000000100` and
the `data` value of `68656c6c6f`.

You can generate your own `namespace_id` and data values using this
useful Golang Playground we created [here](https://go.dev/play/p/ZZWYSSHTRdo).

We run the following:

```sh
curl -X POST -d '{"namespace_id": "0000010000000100", \
  "data": "68656c6c6f", "gas_limit": 100000}' http://127.0.0.1:26658/submit_pfd
```

We get the following output:

```json
{

>    "height":589,

Reviewer 2: we should highlight that this is where the user would find the  block height for which PFD tx was included

   "logs":[
      {
         "events":[
            {
               "type":"message",
               "attributes":[
                  {
                     "key":"action",
                     "value":"/payment.MsgPayForData"
                  }
               ]
            },
            {
               "type":"payfordata",
               "attributes":[
                  {
                     "key":"signer",
                     "value":"celes1vdjkcetnxyc8gandddeh2wfcxe3ksdfn09axkdtrwdnryat3v3jkcut8wpjnjerj0y6hz6sga5fzu"
                  },
                  {
                     "key":"size",
                     "value":"256"
                  }
               ]
            }
         ]
      }
   ],
   "gas_wanted":100000,
   "gas_used":38546
}
```

### Get Namespaced Shares by Block Height

After submitting your PFD transaction, upon success, the node will return
the block height for which the PFD transaction was included. You can then
use that block height and the namespace ID with which you submitted your
PFD transaction to get your message shares returned to you. In this example,
the block height we got was 589 which we will use for the following command.

```sh

> curl -X GET http://127.0.0.1:26658/namespaced_shares/0000000000000001/height/589

Reviewer 2: ```suggestion
curl -X GET http://127.0.0.1:26658/namespaced_shares/0000010000000100/height/589
```

# Getting and Sending Transactions with Celestia Node
<!-- markdownlint-disable MD013 -->

In this tutorial, we will cover how to use the Celestia Node API to submit and
retrieve messages from the Data Availability Layer by their namespace ID.

## Hardware Requirements

You can find hardware requirements [here](../nodes/light-node#hardware-requirements).

## Setting Up Dependencies

You can follow the tutorial for setting up the dependencies [here](../developers/environment).

## Celestia Node

## Install Celestia Node

You can follow the tutorial for building Celestia Node [here](../developers/celestia-node)

### Instantiate Celestia Light Node

Now, let's instantiate a Celestia Light node:

> Note: RPC Endpoints are exposed in all Celestia Node types
  such as Light, Bridge and Full Nodes.

```sh

> ./build/celestia light init

Reviewer 2: in some cases, we do `./build/celestia`, in others we just do `celestia`

lets keep it `celestia` bc we expect users to do make install
